src/libraries/Microsoft.PowerFx.Core/Syntax/Visitors/DefinedTypeDependencyVisitor.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System.Collections.Generic;
+using Microsoft.PowerFx.Core.Binding;
+using Microsoft.PowerFx.Core.Public.Types;
+using Microsoft.PowerFx.Syntax;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Core.Syntax.Visitors
+{
+    internal class DefinedTypeDependencyVisitor : IdentityTexlVisitor
+    {
+        private readonly HashSet<string> _dependencies;
+        private readonly INameResolver _context;
+
+        private DefinedTypeDependencyVisitor(INameResolver context)
+        {
+            _dependencies = new HashSet<string>();
+            _context = context;
+        }
+
+        public static HashSet<string> FindDependencies(TexlNode node, INameResolver context)
+        { 
+            var visitor = new DefinedTypeDependencyVisitor(context);
+            node.Accept(visitor);
+            return visitor._dependencies;
+        }
+
+        public override void Visit(FirstNameNode node)
+        {
+            var name = node.Ident.Name;
+
+            if (_context.LookupType(name, out FormulaType _))
```

Comment:
>_context.LookupType(name, out FormulaType _) [](http://example.com/codeflow?start=16&length=44)

Should this be !_context.LookupType(name, out FormulaType _)?

Reply:
no , if the type for a given name can be looked up in the NameResolver then it is a resolved dependency and hence not adding to the set. I will may invert the logic and comments to make it more underastandable